### PR TITLE
Feature memory-analyzer [depends-on: #2680, #4261]

### PR DESCRIPTION
**Pull request by mmuesly**
The idea of the memory-analyzer, is to create the global state required for some verification harnesses from a core dump file.

In order to do so, you have to feed in a binary with debug symbols and a core file, that matches the binary. And you also have to feed in the static symbols from global state, you care about.

The memory-analyzer will print out c code that is suitable to include it into a test harness in continuation.

This PR won't compile without #2599

It is a clean version of #2648, but it isn't possible to reopen pull request after forced push.

### CODE REVIEW

**Comment by martin-cs** [comment]
Thanks; this is great stuff!  Adding a feature like this might take a little while, just the reviewing, etc. so please bare with us.  Are you working with @tautschnig ?  I know he had an interest in interfacing with GDB.

File: src/ansi-c/c_typecast.cpp
Comment (martin-cs) [inline src/ansi-c/c_typecast.cpp:58]: I'm not sure what these front-end changes have to do with the description of the patch.

File: src/memory-analyzer/gdb_api.cpp
Comment (martin-cs) [inline src/memory-analyzer/gdb_api.cpp:18]: What are the advantages of spawning a GDB process ourselves vs. connecting to a GDB API?  That would allow us to do things like connect to remote machines, connect to other debuggers that implement the same API and to interface with tools like valgrind.

File: src/memory-analyzer/gdb_api.cpp
Comment (martin-cs) [inline src/memory-analyzer/gdb_api.cpp:100]: I'm not personally offended by this kind of thing but it isn't very C++  Would it be possible to use the standard I/O streams to avoid needing to do our own line buffering?

File: src/memory-analyzer/gdb_api.h
Comment (martin-cs) [inline src/memory-analyzer/gdb_api.h:28]: This does worry me a little...

File: src/util/mp_arith.h
Comment (martin-cs) [inline src/util/mp_arith.h:59]: Yes; this commit should definitely get merged.

File: src/memory-analyzer/memory_analyzer_main.cpp
Comment (martin-cs) [inline src/memory-analyzer/memory_analyzer_main.cpp:15]: What are the challenges in making this cross platform?

File: src/memory-analyzer/memory_analyzer_parse_options.cpp
Comment (martin-cs) [inline src/memory-analyzer/memory_analyzer_parse_options.cpp:28]: I don't think we should hardcode the architecture.  Is the usual option of having flags and defaulting to the platform architecture not viable?

File: unit/Makefile
Comment (martin-cs) [inline unit/Makefile]: In future, this kind of unrelated change should probably be in it's own PR.

**Review by martin-cs — CHANGES_REQUESTED** [review]
Generally seems promising and it is definitely a feature we want.  I wonder if we can make this a lot more functional and portable with some small changes.  Also, how does this work with --dump-c ?

File: src/memory-analyzer/memory_analyzer_parse_options.cpp
Comment (mmuesly) [reply src/memory-analyzer/memory_analyzer_parse_options.cpp:28]: I found code that was supposed to read these settings from the goto-binary, but couldn't make that work with my goto binaries.
Anyhow, these information, if passed as a flag, should effect, how gdb is going to interpret the core dump.
Modifying these gdb settings isn't supported right now by the ```gdb_apit```.

So there is more work required than just changing these hard coded values into flags. To prevent users from passing unsupported flags not aligned with the gdb behavior, I hard coded it for now.

File: src/memory-analyzer/memory_analyzer_main.cpp
Comment (mmuesly) [reply src/memory-analyzer/memory_analyzer_main.cpp:15]: GDB is in general not supported out of the box on OSX and Windows. But without GDB, this code doesn't do anything meaningful.

File: src/ansi-c/c_typecast.cpp
Comment (mmuesly) [reply src/ansi-c/c_typecast.cpp:58]: I also introduced another function somewhere in the patch called ```is_void_pointer``` but with other semantics.

The new function checks, that type.id is pointer and subtype.id is void. The older front-end function checks that recursively on subtype. Therefore, I thought renaming the old function is the better choice to resolve the name conflict.

File: src/memory-analyzer/gdb_api.cpp
Comment (mmuesly) [reply src/memory-analyzer/gdb_api.cpp:18]: I have done some research before writing this code, but haven't found a general GDB API. I am happy to reconsider this code and use other solutions.
It seems, like you have a particular API in mind. Might you provide some more details?

File: src/memory-analyzer/gdb_api.h
Comment (mmuesly) [reply src/memory-analyzer/gdb_api.h:28]: Why? Or what are your concerns?

It is just the chunk size, I read from a pipe. The pipe does the real buffering in the background regarding the interprocess communication.
It is certainly interlinked with your comment regarding the stream processing.

**Comment by mmuesly** [comment]
@martin-cs Might you provide some more context on the ```--dump-c``` comment?
Are you concerned, these changes might break the flag?

Most of the changes regarding c dumping required are in another PR: https://github.com/diffblue/cbmc/pull/2599

**Comment by martin-cs** [comment]
@mmuesly ```--dump-c``` : as I understand it, the use-case for this is building calling-contexts / verification harnesses.  One first things I think people will want to do with these is make some parts non-det.  Dumping to C and letting them edit them to their heart's content seems like a good way of doing this, so I was curious to know if the generated code would work with ```--dump-c```.

File: src/ansi-c/c_typecast.cpp
Comment (martin-cs) [reply src/ansi-c/c_typecast.cpp:58]: OK...

File: src/memory-analyzer/gdb_api.cpp
Comment (martin-cs) [reply src/memory-analyzer/gdb_api.cpp:18]: I'm not a GDB expert but:

https://sourceware.org/gdb/current/onlinedocs/gdb/Remote-Debugging.html

makes reference to the "standard GDB remote serial protocol", which I think is the one documented here:

https://sourceware.org/gdb/current/onlinedocs/gdb/Remote-Protocol.html

It seemed like this could be useful for things like this:

https://elinux.org/Debugging_The_Linux_Kernel_Using_Gdb

where you could build calling contexts from a running Linux kernel.

File: src/memory-analyzer/gdb_api.h
Comment (martin-cs) [reply src/memory-analyzer/gdb_api.h:28]: Fixed size buffers make me nervous in general but maybe I'm doing too much security analyses.

File: src/memory-analyzer/memory_analyzer_main.cpp
Comment (martin-cs) [reply src/memory-analyzer/memory_analyzer_main.cpp:15]: This is another argument ( IMHO ) for connecting to a remote GDB rather than spawning our own.

File: src/memory-analyzer/memory_analyzer_parse_options.cpp
Comment (martin-cs) [reply src/memory-analyzer/memory_analyzer_parse_options.cpp:28]: I can see that it makes sense for now but it would be good to avoid this if we can.

File: src/memory-analyzer/gdb_api.cpp
Comment (mmuesly) [reply src/memory-analyzer/gdb_api.cpp:18]: Running the GDB process ourselves keeps this a standalone tool on Linux for now. Connecting to a GDB API requires a network debugger running somewhere. So either we have to take care on setting up the debugger in the network or keep a consistent environment for running the memory analyzer.

I think, that might be on the feature request list, but I don't see, that this is the better solution in general.
Apart from that, I don't want to run a program in the debugger. I mainly want to analyze core dump files at the moment, so there is no benefit to do that in the network as I am not running software inside of gdb at the moment.

I see that this might become desirable for SoC systems. My hope is, that, by separating the gdb_api and the analyzer data structure, given another api later that provides the same functionality, you might exchange the debugger in the background using the same bookkeeping.

Comment (martin-cs) [reply src/memory-analyzer/gdb_api.cpp:18]: I see your point but I think we're going to need someone who actually has merge rights for this to comment.  @tautschnig ?

**Comment by mmuesly** [comment]
I don't see, why this shouldn't work together with --dump-c.
The outputted c code compiles. Therefore, as long as the dump-c code is compilable on its own, you should be able to intermingle results by copy and pasting this into one file.

It is not automagically integrated in a c dump at the moment, if that was your request.

File: src/memory-analyzer/gdb_api.cpp
Comment (mmuesly) [reply src/memory-analyzer/gdb_api.cpp:100]: Yes that is right, but I haven't found a really nice way, to read and write to a child process from C++. Therefore the fork and pipe approach is mainly borrowed from c, but as far as I have seen, there is no standard way, to open a stream buffer on a file descriptor. At least, I haven't found a buffer in I/O stream that opens a file descriptor and searching for this on the internet also hasn't revealed something that works well. For sure there are some Boost libraries doing this for you, but I don't want this dependency.
Instead of reading into this fixed size buffer, I might build some class around the file descriptor and then implement ```>>``` and ```get``` on them. I would then only use ```get```, as I care about whitespaces.

Sure, I might read only a single byte, so I don't maintain an internal buffer, but I see a performance issue here.

I am open for further suggestions and willed to learn, how to change this, but I don't have another idea at hand. Writing my own class with the methods required, doesn't seem to have a benefit from my point of view instead of using this buffer.

**Comment by mmuesly** [comment]
I aligned this PR with #2704. Once, this gets merged, it should be possible make CI happy for this PR as well.

**Comment by mmuesly** [comment]
I adapted the PR to #2713 and fixed some problems with the Make and Make build system. Might someone provide some insight on the Windows Build error? I am willed to fix them as well, once I have understood the problem.

**Comment by smowton** [comment]
@mmuesly please rebase rather than do merge-backs (this may be temporary pending a future rebase in which case ignore me :))

Also if you're not already aware https://github.com/diffblue/cbmc/pull/2816 brings a structured symbol table export format.

File: src/memory-analyzer/Makefile
Comment (tautschnig) [inline src/memory-analyzer/Makefile]: Do not hard-code `.a`, use `$(LIBEXT)` as done elsewhere - this is currently causing build failures on Windows.

**Comment by tautschnig** [comment]
I have now marked this as depends-on #4261 just for book keeping purposes - my understanding of #4261 is that @danpoe has picked up all the commits from this PR (or at least so it seemed to me as most commits in #4261 are co-authored by @mmuesly). @danpoe can you confirm?

**Comment by tautschnig** [comment]
Closing as the patches have been merged in #4261.

**Pull request closed without merging**
